vllm/spec_decode/draft_model_runner.py

```diff
@@ -115,8 +120,228 @@ def update_model_input(
+        self._update_sampling_metadata(sampling_metadata, num_seqs,
+                                       num_queries)
+
+        # Create new input
+        new_model_input = self._model_input_cls(
+            input_tokens=model_input.input_tokens,
+            input_positions=model_input.input_positions,
+            attn_metadata=attn_metadata,
+            seq_lens=attn_metadata.seq_lens,
+            query_lens=model_input.query_lens,
+            lora_mapping=model_input.lora_mapping,
+            lora_requests=model_input.lora_requests,
+            multi_modal_kwargs=model_input.multi_modal_kwargs,
+            sampling_metadata=model_input.sampling_metadata,
+            is_prompt=False,
+        )
+
+        if log_advance_input:
+            print("NEW INPUT: ")
+            print("  input_tokens = {}".format(new_model_input.input_tokens))
+            print("  input_positions = {}".format(
+                new_model_input.input_positions))
+            print("  seq_lens = {}".format(new_model_input.seq_lens))
+            print("  query_lens = {}".format(new_model_input.query_lens))
+            print("  attn_metadata:")
+            print("    seq_lens_tensor: {}".format(
+                attn_metadata.seq_lens_tensor))
+            print("    slot_mapping: {}".format(attn_metadata.slot_mapping))
+            print("    block_tables: {}".format(attn_metadata.block_tables))
+
+        return new_model_input
+
+    def _can_use_advance_step(self):
+        if not enable_advance_step:
+            return False
+
+        # TODO: Add support for other attn backends
+        if self.attn_backend.get_name() != "flash-attn":
+            return False
+
+        # TODO: Understand what is needed for LORA support
+        if self.lora_config:
+            return False
+
+        return True
+
+    def _get_model(self, model_input):
+        # TODO: Expand to more backends
+        assert model_input.attn_metadata is not None
+        assert isinstance(model_input.attn_metadata, FlashAttentionMetadata)
+
+        # Currently cuda graph is only supported by the decode phase.
+        is_decode = (model_input.attn_metadata.num_prefills == 0
+                     and model_input.attn_metadata.use_cuda_graph)
+        if is_decode:
+            assert model_input.input_tokens is not None
+            graph_batch_size = model_input.input_tokens.shape[0]
+            model_executable = self.graph_runners[graph_batch_size]
+        else:
+            model_executable = self.model
+
+        return model_executable
+
+    @torch.inference_mode()
+    def _execute_model_with_advance_step(
+            self, model_input: ModelInputForGPUWithSamplingMetadata,
+            kv_caches: List[torch.Tensor],
+            num_steps: int) -> Optional[List[SamplerOutput]]:
+
+        outputs: List[SamplerOutput] = []
+        for step in range(num_steps):
+            model_executable = self._get_model(model_input)
```

Comment:
Can we reuse the same executable so that we don't need to run this function every time? I suppose even in the worst case that step-0 is prefill, we can still reuse the executable starting from step-1.

Reply:
Good idea. I have restricted the draft_model_runner multi-step to decodes only, so its not an issue anymore.